app/controllers/spree/digitals_controller.rb

```diff
@@ -2,42 +2,29 @@ module Spree
   class DigitalsController < Spree::StoreController
     force_ssl only: :show, if: :ssl_configured?
     rescue_from ActiveRecord::RecordNotFound, with: :resource_not_found
+    before_action :authorize_digital_link, only: :show
 
     def show
-      if attachment.present?
-        # don't authorize the link unless the file exists
-        # the logger error will help track down customer issues easier
-        if attachment_is_file?
-          if digital_link.authorize!
-            if Paperclip::Attachment.default_options[:storage] == :s3
-              redirect_to attachment.expiring_url(Spree::DigitalConfiguration[:s3_expiration_seconds]) and return
-            else
-              send_file attachment.path, filename: attachment.original_filename, type: attachment.content_type and return
-            end
-          end
-        else
-          Rails.logger.error "Missing Digital Item: #{attachment.path}"
-        end
+      if digital_link.cloud?
+        redirect_to attachment.expiring_url(Spree::DigitalConfiguration[:s3_expiration_seconds])
+      else
+        send_file attachment.path, filename: attachment.original_filename, type: attachment.content_type
       end
-
-      render :unauthorized
     end
 
     private
-      def attachment_is_file?
-        if Paperclip::Attachment.default_options[:storage] == :s3
-          attachment.exists?
-        else
-          File.file?(attachment.path)
-        end
+      def authorize_digital_link
+        # don't authorize the link unless the file exists
+        raise ActiveRecord::RecordNotFound unless attachment.present?
+        render :unauthorized unless digital_link.is_file? && digital_link.authorize!
```

Comment:
For me `unless xx && yy` is super hard to understand (I have to go through it several times in my head to understand what is means. I understand `if !xx || !yy` much quicker. But maybe that's just me, so it's your call.